Core/src/Surfaces/CurvilinearSurface.cpp

```diff
@@ -0,0 +1,132 @@
+  Transform3 transform = Transform3(referenceFrame());
+  transform.pretranslate(center());
+  return transform;
+}
+
+/// Calculate the jacobian from local to global which the surface knows best,
+/// hence the calculation is done here.
+///
+/// @return Jacobian from local to global
+BoundToFreeMatrix CurvilinearSurface::boundToFreeJacobian() const {
+  // Prepare the jacobian to free
+  BoundToFreeMatrix jacobian = BoundToFreeMatrix::Zero();
+
+  auto [cosPhi, sinPhi, cosTheta, sinTheta] =
+      VectorHelpers::evaluateTrigonomics(m_direction);
+  jacobian(eFreePos0, eBoundLoc0) = -sinPhi;
+  jacobian(eFreePos0, eBoundLoc1) = -cosPhi * cosTheta;
+  jacobian(eFreePos1, eBoundLoc0) = cosPhi;
+  jacobian(eFreePos1, eBoundLoc1) = -sinPhi * cosTheta;
+  jacobian(eFreePos2, eBoundLoc1) = sinTheta;
+  // Time parameter: stays as is
+  jacobian(eFreeTime, eBoundTime) = 1;
+  jacobian.block<3, 2>(eFreeDir0, eBoundPhi) =
+      sphericalToFreeDirectionJacobian(m_direction);
+  // Q/P parameter: stays as is
+  jacobian(eFreeQOverP, eBoundQOverP) = 1;
+
+  return jacobian;
+}
+
+/// Calculate the jacobian from global to local which the surface knows best,
+/// hence the calculation is done here.
+///
+/// @return Jacobian from global to local
+FreeToBoundMatrix CurvilinearSurface::freeToBoundJacobian() const {
+  // Prepare the jacobian to curvilinear
+  FreeToBoundMatrix jacobian = FreeToBoundMatrix::Zero();
+
+  auto [cosPhi, sinPhi, cosTheta, sinTheta] =
+      VectorHelpers::evaluateTrigonomics(m_direction);
+  // We operate in curvilinear coordinates defined as follows
+  jacobian(eBoundLoc0, eFreePos0) = -sinPhi;
+  jacobian(eBoundLoc0, eFreePos1) = cosPhi;
+  jacobian(eBoundLoc1, eFreePos0) = -cosPhi * cosTheta;
+  jacobian(eBoundLoc1, eFreePos1) = -sinPhi * cosTheta;
+  jacobian(eBoundLoc1, eFreePos2) = sinTheta;
+  // Time parameter: stays as is
+  jacobian(eBoundTime, eFreeTime) = 1.;
+  // Directional and momentum parameters for curvilinear
+  jacobian.block<2, 3>(eBoundPhi, eFreeDir0) =
+      freeToSphericalDirectionJacobian(m_direction);
+  // Q/P parameter: stays as is
+  jacobian(eBoundQOverP, eFreeQOverP) = 1.;
+
+  return jacobian;
+}
+
+FreeToPathMatrix CurvilinearSurface::freeToPathDerivative() const {
+  FreeToPathMatrix freeToPath = FreeToPathMatrix::Zero();
+  freeToPath.segment<3>(eFreePos0) = -1.0 * m_direction;
+  return freeToPath;
+}
+
+std::ostream& CurvilinearSurface::toStream(std::ostream& sl) const {
+  sl << std::setiosflags(std::ios::fixed);
+  sl << std::setprecision(4);
+  sl << std::boolalpha;
+  sl << "Curvilinear Surface" << std::endl;
+  sl << "     Center position  (x, y, z) = (" << m_position.x() << ", "
+     << m_position.y() << ", " << m_position.z() << ")" << std::endl;
+  sl << "     Direction  (x, y, z) = (" << m_direction.x() << ", "
+     << m_direction.y() << ", " << m_direction.z() << ")" << std::endl;
+  sl << std::setprecision(-1);
+  sl << std::noboolalpha;
```

Comment:
I'm going to be **very** pedantic here... :thinking: This is the type of code that makes/made stream based printouts a pain in C\+\+. :frowning: That one has to be very careful not to change the global state of the stream passed to a function like this one.

In this case the received stream can very much walk away from this function in a "different state" than it arrived with. :frowning: The most pragmatic solution to which is to:
  - Construct the formatted output using an `std::ostringstream`;
  - Put the contents of that stream into the stream passed to this function.

I.e. you might as well remove this function, and just provide the `toString()` function. (Implemented on its own this time around.) Then you don't need to worry about leaking stream modifiers. :thinking:

If you do that, of course the "global" output operator would become:

```c++
std::ostream& operator<<(std::ostream& os, const CurvilinearSurface& surface) {
  out << surface.toSting();
  return out;
}
```

Reply:
Right, I was also worried about changing the state of the streamer but went blind since I just copied the code from https://github.com/acts-project/acts/blob/b0c64a07b4603bad9e9df94bf0b369bbbf795ae8/Core/src/Surfaces/Surface.cpp#L198-L221 (looks not exactly like this so might actually have been a different source) and wanted to be done with it 😄 

Using a string would be one option. Restoring the stream to the original state another I guess? Do you know if there are convenience functions for that or do we need to remember the properties we change before setting them?

I will open an issue about this.

Reply:
https://github.com/acts-project/acts/issues/3079

Reply:
It wouldn't be impossible to use a combination of:
  - https://en.cppreference.com/w/cpp/io/ios_base/flags
  - https://en.cppreference.com/w/cpp/io/ios_base/setf
  - https://en.cppreference.com/w/cpp/io/ios_base/unsetf

But the last time I tried this, I gave up in the end. As the code became just too finicky. It was just easier to create a temporary stream whose flags I could modify however I wanted to. :thinking: